cip/CIP2015-10-03-State-visibility-between-clauses.adoc

```diff
@@ -0,0 +1,76 @@
+= CIP20140109 - State visibility between clauses
+:numbered:
+:toc:
+:toc-placement: macro
+:source-highlighter: codemirror
+
+*Authors:* Andrés Taylor <andres@neotechnology.com>
+
+[abstract]
+.Abstract
+--
+This CIP tries to formalise how state changes behave while inside a Cypher query. Currently, because this is not explicit enough, the behaviour is implementation dependent.
+--
+
+toc::[]
+
+=== Background
+Cypher allows queries to go through multiple steps, and it’s possible to interleave reading and writing to the graph.
+Some clauses can also do both - read the graph and write to it at the same time.
+
+To allow the user to understand what the outcome will be at an abstract, logical level, semantics for the visibility of these changes should be clear.
+
+=== Proposal
+Each clause lives in its own state, which includes all the changes of clauses coming before it, including changes performed by itself, *but none of the changes made by clauses listed later in the query.*
+
+The semantical data flow here is such that each clause operates on the entire result set before passing control to any subsequent clause.
+
+Updates between UNION subqueries are also treated in a sequential manner.
+Therefore subqueries coming later will see updates from subqueries before, but not the other way around.
+
```

Comment:
I think all keywords ought to be formatted, so please change "UNION" -> "`UNION`"